x/tokenomics/keeper/settle_pending_claims.go

```diff
@@ -32,45 +32,49 @@ func (k Keeper) SettlePendingClaims(ctx cosmostypes.Context) (
 	logger := k.Logger().With("method", "SettlePendingClaims")
 	logger.Debug("settling expiring claims")
 
-	// Retrieve all expiring claims as an iterator.
-	// DEV_NOTE: This previously retrieve a list but has been change to account for large claim counts.
-	expiringClaimsIterator := k.GetExpiringClaimsIterator(ctx)
-	defer expiringClaimsIterator.Close()
-
-	blockHeight := ctx.BlockHeight()
-
 	// Initialize results structs.
+	blockHeight := ctx.BlockHeight()
 	settledResults = make(tlm.ClaimSettlementResults, 0)
 	expiredResults = make(tlm.ClaimSettlementResults, 0)
 	settlementContext := NewSettlementContext(ctx, &k, logger)
 	numExpiringClaims := 0
 	numDiscardedFaultyClaims = 0
 
+	// Retrieve an iterator of all expiring claims.
+	// DEV_NOTE: This previously retrieve a list but has been change to account for large claim counts.
+	expiringClaimsIterator := k.GetExpiringClaimsIterator(ctx, settlementContext, blockHeight)
+	defer expiringClaimsIterator.Close()
+
 	// Iterating over all potentially expiring claims.
 	// This loop does the following:
-	// - Retrieve the claim
-	// - Settle the claim
-	// - Remove the claim from the state
-	// - Emit an event
-	// - Update the relevant actors in the state
+	// 1. Retrieve the claim
+	// 2. Settle the claim
+	// 3. Remove the claim from the state
+	// 4. Emit an event
+	// 5. Update the relevant actors in the state
 	for ; expiringClaimsIterator.Valid(); expiringClaimsIterator.Next() {
 		claim, iterErr := expiringClaimsIterator.Value()
 		if iterErr != nil {
-			claimKey := string(expiringClaimsIterator.Key())
-			logger.Error(tokenomicstypes.ErrTokenomicsSettlementInternal.Wrapf(
-				"[UNEXPECTED ERROR] Critical error during claim settlement during session with key %s. Claim will be discarded to prevent chain halt: %s",
-				claimKey, err,
-			).Error())
 			numDiscardedFaultyClaims++
+			claimKey := string(expiringClaimsIterator.Key())
+			claimErr := tokenomicstypes.ErrTokenomicsSettlementInternal.Wrapf(
+				"[UNEXPECTED ERROR] Critical error during claim settlement during session with key %s. Claim will be discarded to prevent chain halt: %v",
+				claimKey, iterErr)
+			logger.Error(claimErr.Error())
 			continue
 		}
 
 		// Settle the claim.
 		claimProcessingContext, settlementErr := k.settleClaim(ctx, settlementContext, claim, logger)
 		if settlementErr != nil {
 			// Discard a faulty claim and continue iterating over the next one.
-			k.discardFaultyClaim(ctx, claim, settlementErr, logger)
 			numDiscardedFaultyClaims++
+			err = tokenomicstypes.ErrTokenomicsSettlementInternal.Wrapf(
+				"[UNEXPECTED ERROR] Critical error during claim settlement during session %q for operator %s and service %s. Claim will be discarded to prevent chain halt: %v",
```

Comment:
```suggestion
				"[UNEXPECTED ERROR] Critical error during claim settlement during session %q for operator %s and service %s. Claim will be discarded to prevent chain halt: %s",
```